src/optimizer/stats/cost.cpp

```diff
@@ -202,8 +241,303 @@ double Cost::OrderByCost(const std::shared_ptr<TableStats> &input_stats,
+  UpdateJoinOutputSize(left_input_stats, right_input_stats, output_stats,
+                       predicate, 0);
+  return left_input_stats->num_rows * right_input_stats->num_rows *
+         DEFAULT_TUPLE_COST;
+}
+
+//===----------------------------------------------------------------------===//
+// LEFT NL JOIN
+//===----------------------------------------------------------------------===//
+double Cost::LeftNLJoinWithSampling(
+    const std::shared_ptr<TableStats> &left_input_stats,
+    const std::shared_ptr<TableStats> &right_input_stats,
+    std::shared_ptr<TableStats> &output_stats,
+    const std::shared_ptr<expression::AbstractExpression> predicate) {
+  UpdateJoinOutputSize(left_input_stats, right_input_stats, output_stats,
+                       predicate, left_input_stats->num_rows);
+  return left_input_stats->num_rows * right_input_stats->num_rows *
+         DEFAULT_TUPLE_COST;
+}
+
+//===----------------------------------------------------------------------===//
+// RIGHT NL JOIN
+//===----------------------------------------------------------------------===//
+double Cost::RightNLJoinWithSampling(
+    const std::shared_ptr<TableStats> &left_input_stats,
+    const std::shared_ptr<TableStats> &right_input_stats,
+    std::shared_ptr<TableStats> &output_stats,
+    const std::shared_ptr<expression::AbstractExpression> predicate) {
+  UpdateJoinOutputSize(left_input_stats, right_input_stats, output_stats,
+                       predicate, right_input_stats->num_rows);
+  return left_input_stats->num_rows * right_input_stats->num_rows *
+         DEFAULT_TUPLE_COST;
+}
+
+//===----------------------------------------------------------------------===//
+// OUTER NL JOIN
+//===----------------------------------------------------------------------===//
+double Cost::OuterNLJoinWithSampling(
+    const std::shared_ptr<TableStats> &left_input_stats,
+    const std::shared_ptr<TableStats> &right_input_stats,
+    std::shared_ptr<TableStats> &output_stats,
+    const std::shared_ptr<expression::AbstractExpression> predicate) {
+  UpdateJoinOutputSize(
+      left_input_stats, right_input_stats, output_stats, predicate,
+      left_input_stats->num_rows + right_input_stats->num_rows);
+  return left_input_stats->num_rows * right_input_stats->num_rows *
+         DEFAULT_TUPLE_COST;
+}
+
+//===----------------------------------------------------------------------===//
+// INNER HASH JOIN
+//===----------------------------------------------------------------------===//
+double Cost::InnerHashJoinWithSampling(
+    const std::shared_ptr<TableStats> &left_input_stats,
+    const std::shared_ptr<TableStats> &right_input_stats,
+    std::shared_ptr<TableStats> &output_stats,
+    const std::shared_ptr<expression::AbstractExpression> predicate) {
+  UpdateJoinOutputSize(left_input_stats, right_input_stats, output_stats,
+                       predicate, 0);
+  return (left_input_stats->num_rows + right_input_stats->num_rows) *
```

Comment:
Hash join cost doesn't seem right. There are costs of building a hash table (left table or right table depending on the join type) and performing the actual join.